### PR TITLE
RDCC-2391: Delete API - Delete Users by Email Pattern or User ID

### PR DESCRIPTION
**Pull request by AdilOoze**
### JIRA link (if applicable) ###

[RDCC-2391](https://tools.hmcts.net/jira/browse/RDCC-2391)

### Change description ###

Enhancing Delete API to Delete Users by Email Pattern or User ID

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java [hunk 1]
```diff
@@ -30,16 +37,66 @@
...
+        List<UserProfile> userProfiles = userProfileRepository.findByEmailIgnoreCaseContaining(emailPattern);
+
+        if (userProfiles.isEmpty()) {
+            throw new ResourceNotFoundException("No User Profiles found for email pattern: " + emailPattern);
+        }
+
+        for (int i = userProfiles.size() - 1; i >= 0; i--) {
+            Response idamResponse = idamClient.deleteUser(userProfiles.get(i).getIdamId());
+
+            if (idamResponse.status() == NO_CONTENT.value() || idamResponse.status() == NOT_FOUND.value()) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:76]: NO_CONTENT is success scenario for delete from Idam ?

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java [hunk 2]
```diff
@@ -30,16 +37,66 @@
...
+                break;
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:77]: Correct me if I am wrong here. why are we breaking out of the loop if there is a success response?

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java (hunk 1 above)
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:76]: Yes `NO_CONTENT` with `204` status is what is returned for successful deletion in IDAM.

_Please see their Swagger documentation [here](https://idam-api.aat.platform.hmcts.net/swagger-ui.html#/User_Role_Management)_

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java (hunk 2 above)
Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:77]: If the response is not successful then we do not want to delete them from UP so we remove them from the `userProfiles` list. This list is sent to the `deleteUserProfiles` method on line 89 for deletion in UP.

However, if the response is successful, then we do not have to remove them from the list and therefore we continue our loop.

Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java]: This has been addressed here: https://github.com/hmcts/rd-user-profile-api/pull/631/commits/da42b45bd8eb176b75fd12404b933ecd4a74f7af - thanks for your help @SujeethKumarAri 😄

**Review by SujeethKumarAri — DISMISSED** [review]

**Review by shreedhar-hmcts — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/exception/ForbiddenException.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package uk.gov.hmcts.reform.userprofileapi.exception;
+
+public class ForbiddenException extends RuntimeException {
+
+    public ForbiddenException(String message) {
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/exception/ForbiddenException.java:5]: Use Single exception. May be create generic UserPRofileFailureException and Pass message Forbidden LD

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/FeatureToggleServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    public FeatureToggleServiceImpl(LDClient ldClient, @Value("${launchdarkly.sdk.user}") String userName) {
+        this.ldClient = ldClient;
+        this.userName = userName;
+    }
+
+    @PostConstruct
+    public void mapServiceToFlag() {
+        launchDarklyMap = new HashMap<>();
+        launchDarklyMap.put("UserProfileController.deleteUserProfileByIdOrEmailPattern",
+                "up-delete-by-id-or-email-pattern");
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/FeatureToggleServiceImpl.java:37]: Please verify flag name with Shashank

Comment (AdilOoze) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/FeatureToggleServiceImpl.java]: @shashank-rastogi has confirmed the flag name to be `delete-user-by-id-or-email-pattern`, change made here: https://github.com/hmcts/rd-user-profile-api/pull/631/commits/dc8401546b54112d1f427da53ee4204c78eb8a92

**Review by abhijitdiwan — APPROVED** [review]

**Review by abhijitdiwan — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java [hunk 5]
```diff
@@ -30,16 +36,61 @@
...
     @Value("${loggingComponentName}")
     private String loggingComponentName;
 
+    @Override
+    @Transactional
+    public UserProfilesDeletionResponse deleteByUserId(String userId) {
+        Response idamResponse = idamClient.deleteUser(userId);
+
+        if (idamResponse.status() == NO_CONTENT.value() || idamResponse.status() == NOT_FOUND.value()) {
+            UserProfile userProfile = validateUserStatus(userId);
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:51]: should n't validateUserStatus() method be called before deleting user from idam?

how does it handle scenario where user exists in SIDAM but not in UP?

Comment (shashank-rastogi) [reply src/main/java/uk/gov/hmcts/reform/userprofileapi/service/impl/DeleteUserProfileServiceImpl.java:51]: may not necessarily be a review comment but i need to understand how the flow will work in this case?

**Pull request closed without merging**